### PR TITLE
feat: add a specialized writeStream implementation for s3 external storage

### PR DESCRIPTION
**Pull request by icewind1991**
This should both reduce overhead and improve detection of failed writes

### CODE REVIEW

File: apps/files_external/lib/Lib/Storage/AmazonS3.php [hunk 1]
```diff
@@ -754,4 +755,23 @@ public function hasUpdated($path, $time) {
 			return true;
 		}
 	}
+
+	public function writeStream(string $path, $stream, ?int $size = null): int {
+		if ($size === null) {
+			$size = 0;
+			$stream = CountWrapper::wrap($stream, function (int $writtenSize) use (&$size) {
+				$size = $writtenSize;
+			});
```
Comment (artonge) [inline apps/files_external/lib/Lib/Storage/AmazonS3.php:764]: Is it a way to get the final written size? This is triggered on `writeObject`, right? Can you add a comment explaining it?

Comment (icewind1991) [reply apps/files_external/lib/Lib/Storage/AmazonS3.php:765]: done

**Review by artonge — APPROVED** [review]

**Review by sorbaugh — APPROVED** [review]

**Pull request merged**